### PR TITLE
refactor(clone): Creates CloneVisitor and CloneBuilder.

### LINKED ISSUE #662: add contract check equals(e,clone(e)) == true

### PR DESCRIPTION
**Pull request by GerardPaligot**
Closes #662

### CODE REVIEW

File: src/test/java/spoon/test/main/MainTest.java [hunk 1]
```diff
@@ -79,6 +80,24 @@ public void checkGenericContracts(CtPackage pack) {
...
+		new CtBiScannerDefault() {
+			@Override
+			public boolean biScan(CtElement element, CtElement other) {
+				if (element != null || other != null) {
+					assertNotNull(element);
+					assertNotNull(other);
+					assertEquals(element, other);
+					assertFalse(element == other);
+				}
+				return super.biScan(element, other);
```
Comment (monperrus) [inline src/test/java/spoon/test/main/MainTest.java]: what does the super implementation?

File: src/test/java/spoon/test/main/MainTest.java [hunk 2]
```diff
@@ -79,6 +80,24 @@ public void checkGenericContracts(CtPackage pack) {
...
+
+		// clone
+		checkEqualityBetweenOriginalAndClone(pack);
+	}
+
+	private void checkEqualityBetweenOriginalAndClone(CtPackage pack) {
+		new CtBiScannerDefault() {
+			@Override
+			public boolean biScan(CtElement element, CtElement other) {
+				if (element != null || other != null) {
```
Comment (monperrus) [inline src/test/java/spoon/test/main/MainTest.java]: heck required? can this happen that both are null?

**Comment by monperrus** [comment]
Where is the change that switches from the old clone code to the new one? where do we remove the old one?

File: src/test/java/spoon/test/main/MainTest.java (hunk 1 above)
Comment (GerardPaligot) [inline src/test/java/spoon/test/main/MainTest.java]: The same thing that `scan` in `CtScanner`.

File: src/test/java/spoon/test/main/MainTest.java (hunk 2 above)
Comment (GerardPaligot) [inline src/test/java/spoon/test/main/MainTest.java]: No, so I check if there aren't null just after.

**Comment by GerardPaligot** [comment]
Here: https://github.com/INRIA/spoon/pull/717/files#diff-ab96d7e6a8e0811ebd1b2f78dcfab024R458

**Comment by monperrus** [comment]
Why don't we change Core().clone() as well?

File: src/main/java/spoon/generating/clone/CloneVisitorTemplate.java [hunk 3]
```diff
@@ -0,0 +1,93 @@
...
+import java.util.Set;
+import java.util.TreeSet;
+
+/**
+ * Used to clone a given element.
+ *
+ * This class is generated automatically by the processor {@link spoon.generating.CloneVisitorGenerator}.
+ */
+class CloneVisitorTemplate extends CtScanner {
+	public static <T extends CtElement> T clone(T element) {
```
Comment (GerardPaligot) [inline src/main/java/spoon/generating/clone/CloneVisitorTemplate.java]: Move to helper class.

File: src/main/java/spoon/support/reflect/code/CtLambdaImpl.java [hunk 4]
```diff
@@ -66,7 +66,7 @@ public String getSimpleName() {
 
 	@Override
 	public <B extends T, C extends CtExecutable<T>> C setBody(CtBlock<B> body) {
```
Comment (GerardPaligot) [inline src/main/java/spoon/support/reflect/code/CtLambdaImpl.java]: Think to check architecture test.

File: src/main/java/spoon/support/visitor/clone/CloneBuilder.java [hunk 5]
```diff
@@ -0,0 +1,231 @@
...
+		return ((T) (builder.other));
+	}
+
+	private spoon.reflect.declaration.CtElement other;
+
+	public void setOther(spoon.reflect.declaration.CtElement other) {
+		spoon.support.visitor.clone.CloneBuilder.this.other = other;
+	}
+
+	public <T> void visitCtCodeSnippetExpression(spoon.reflect.code.CtCodeSnippetExpression<T> e) {
```
Comment (GerardPaligot) [inline src/main/java/spoon/support/visitor/clone/CloneBuilder.java]: Inline comment "auto-generated see <Processor>"

File: src/test/java/spoon/test/main/MainTest.java (hunk 2 above)
Comment (GerardPaligot) [inline src/test/java/spoon/test/main/MainTest.java]: Think to check not null.

File: src/test/java/spoon/test/main/MainTest.java [hunk 6]
```diff
@@ -79,6 +80,24 @@ public void checkGenericContracts(CtPackage pack) {
 
 		// scanners
 		checkContractCtScanner(pack);
+
+		// clone
+		checkEqualityBetweenOriginalAndClone(pack);
+	}
+
+	private void checkEqualityBetweenOriginalAndClone(CtPackage pack) {
```
Comment (GerardPaligot) [inline src/test/java/spoon/test/main/MainTest.java]: Adds scanner counter and check that is equals with BiScanner.

**Comment by GerardPaligot** [comment]
Think to remove deprecation on old clone and remove old implementation.

File: src/test/java/spoon/test/main/MainTest.java (hunk 6 above)
Comment (GerardPaligot) [inline src/test/java/spoon/test/main/MainTest.java]: Done

**Comment by GerardPaligot** [comment]
@monperrus PR ok for me.

**Pull request merged**